Cache channels before updating server data, fixes #374

z64 opened the pull request:
This fixes a bug where channels we don't have access to (because we can't Read Messages or Connect) are attempted to be resolved by `GET /channels/{cid}`, when in fact, we already have the data for these channels in `GUILD_CREATE` and can properly cache them from there.

For example, this would cause exceptions to be raised every time a user joins or leaves (voice state update) a voice channel that denies Connect to our client.

For more information on this behavior, see https://github.com/meew0/discordrb/issues/374#issuecomment-313839853

The logic in `update_data` that refers to a channel being "unavailable" and setting AFK channels to nil can probably be removed, and maybe the embed_channel as well, but I need to look into this more closely to tidy it up. Opening this PR as a placeholder in the meantime.

---

Todo:
- [x] Move `update_data` back up, but place `process_channels` before it to isolate any future edge cases
- [x] Remove "unavailable channel" exception check for `@afk_channel`
- [x] Remove "unavailable channel" exception check for `@embed_channel`
- [x] Fix `||` logic in some `update_data` statements

meew0 commented:
Are you sure this doesn't cause any problems with some of the other `process_` methods requiring server data to be present? From a quick look I haven't seen any problems but there's a lot of code in e.g. initializers I might have overlooked.

z64 commented:
As we discussed in Discord I moved the location of `process_channels` / `update_data`, and removed the "unavailability" handling that should never occur.

Also, see my comments in the full commit message of 2baf63a. I updated behavior of `Server#update_data` to remove the need for `cache_embed`. Just to recap, it was necessary to resolve the server as `embed_channel_id`, as `GUILD_CREATE` doesn't provide `embed_channel_id`, and we may have not otherwise received an update yet to the guild that sets that attribute. We were just picking out that attribute, but we may as well not "waste" the rest of the "fresh" data.

I tested all of the immediately affected methods and everything works fine. I may do some more testing later (leaving my bot on this branch for now too).

z64 commented:
Before I forget too, I'll also fix the third clause in the `update_data` statements from not being used due to the second clause not being nilable in this PR, as we discussed briefly

z64 commented:
My todo's are done here.

z64 commented:
I'll note from 9e92f76, in a later update, we should instead use the REST calls specifically for guild embeds ( #435 ) instead of the whole server object.
